data/trie/patriciaMerkleTrie.go

```diff
@@ -413,258 +383,87 @@ func (tr *patriciaMerkleTrie) Prune(rootHash []byte, identifier data.TriePruning
 	defer tr.mutOperation.Unlock()
 
 	rootHash = append(rootHash, byte(identifier))
-
-	if tr.snapshotInProgress {
-		tr.pruningBuffer = append(tr.pruningBuffer, rootHash)
-		return nil
-	}
-
-	err := tr.removeFromDb(rootHash)
-	if err != nil {
-		return err
-	}
-
-	return nil
+	return tr.trieStorage.Prune(rootHash)
 }
 
 // CancelPrune invalidates the hashes that correspond to the given root hash from the eviction waiting list
 func (tr *patriciaMerkleTrie) CancelPrune(rootHash []byte, identifier data.TriePruningIdentifier) {
 	tr.mutOperation.Lock()
 	rootHash = append(rootHash, byte(identifier))
-	_, _ = tr.dbEvictionWaitingList.Evict(rootHash)
+	tr.trieStorage.CancelPrune(rootHash)
 	tr.mutOperation.Unlock()
```

Comment:
Done.

Reply:
I disagree. We should use `defer unlock` only in cases where we have multiple "exit points" inside a function. Otherwise, we can unlock after the usage of the protected resource. Defer will involve a little bit more computation and if it is not necessary we can avoid it.

Reply:
For me it's ok either way. @ccorcoveanu ?

Reply:
@bogdan-rosianu Yes, that's the general recommendation using mutexes. But in this special case it seems the use case is to lock the mutex through the whole function execution. It's not as performant but it's a lot easier to read. 
Consider that somebody will edit the function later in time adding some functionality. He will have to go through the whole thought process to figure out if the new code should or shouldn't be protected

Reply:
If you still don't agree we can leave it as it was :) I don't have a strong opinion, I just prefer code that is easy to read :)

Reply:
I understand what you say, but in this case the functions were pretty simple. In case of larger functions, we can use defers. Also, this is inside the trie so we can use each small improvement. So @BeniaminDrasovean , please put it as it was before

Reply:
Exactly, I was talking about those simple things.
```go
func (tr *patriciaMerkleTrie) AppendToOldHashes(hashes [][]byte) {
	tr.mutOperation.RLock()
	tr.oldHashes = append(tr.oldHashes, hashes...)
        tr.mutOperation.RUnlock()

        // ^ This is a classic lock - And now, what if I do:
       tr.trieStorage.TakeSnapshot(rootHash, tr.marshalizer, tr.hasher)
}
```

Reply:
Changed with `Lock()` and `Unlock()` instead of `RLock()` and `RUnlock()`

Reply:
@iulianpascalau please leave an opinion

Reply:
My opinion is to use mut.Lock and mut.Unlock (without defer) if the function is concise with no early exits. In this way we have a slight faster execution times. Moreover, since we are dealing with tries, I suggest to keep the code as optimum as we can.

Reply:
I don't like it, but ok :)